### PR TITLE
RAV-6507 Support of REGEXP_SUBSTR function at BigQuery End

### PR DESCRIPTION
**Pull request by yogeshmuley**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 1]
```diff
@@ -584,18 +583,17 @@ private SqlNode makCastCallForTimeWithPrecision(SqlNode operandToCast, int preci
   private void unparseRegexSubstr(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
     SqlCall extractCall;
     switch (call.operandCount()) {
-    case 3:
+    case 2: case 3:
```
Comment (soma-mondal) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:586]: Can we remove the switch case as both 2,3 and 4,5 operand count have same code now.

Comment (ankitaspanse) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: I have added the changes as per the review comments as I observed no offset is been added at bigquery end as per examples provide at bigquery side.

**Review by soma-mondal — APPROVED** [review]

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 2]
```diff
@@ -583,44 +582,32 @@ private SqlNode makCastCallForTimeWithPrecision(SqlNode operandToCast, int preci
...
-    regexStr = regexStr.replace("\\", "\\\\");
-    String regexLiteral = "(?i)".concat(regexStr.substring(1, regexStr.length() - 1));
-    return SqlLiteral.createCharString(regexLiteral,
-        call.operand(1).getParserPosition());
+    String regexLiteral = call.operand(1).toString();
+    regexLiteral = regexLiteral.replace("\\", "\\\\");
+    if (call.operandCount() == 5 && call.operand(4).toString().equals("'i'")) {
+      regexLiteral = "(?i)".concat(regexLiteral.substring(1, regexLiteral.length() - 1));
+    }
+    return SqlLiteral.createCharString(regexLiteral, call.operand(1).getParserPosition());
```
Comment (Technay) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:610]: substring just once while returning

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 3]
```diff
@@ -5859,18 +5858,17 @@ private void checkLiteral2(String expression, String expected) {
...
         + "from \"foodmart\".\"product\" where \"product_id\" in (1, 2, 3, 4)";
     final String expected = "SELECT "
-        + "REGEXP_EXTRACT_ALL(SUBSTR('chocolate Chip cookies', 1), '(?i)[-\\\\_] V[0-9]+') [OFFSET(0)]\n"
+        + "REGEXP_SUBSTR('chocolate Chip cookies', '(?i)[-\\\\_] V[0-9]+', 1, 1)\n"
         + "FROM foodmart.product\n"
         + "WHERE product_id = 1 OR product_id = 2 OR product_id = 3 OR product_id = 4";
     sql(query)
         .withBigQuery()
         .ok(expected);
   }
```
Comment (Technay) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:5870]: check whether you can pass string arguments in 3rd/4th parameter

**Review by Technay — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 4]
```diff
@@ -583,44 +582,32 @@ private SqlNode makCastCallForTimeWithPrecision(SqlNode operandToCast, int preci
...
-    writer.literal("[OFFSET(" + occurrenceNumber + ")]");
+    extractCall = makeExtractSqlCall(call);
+    REGEXP_SUBSTR.unparse(writer, extractCall, leftPrec, rightPrec);
   }
 
   private SqlCall makeExtractSqlCall(SqlCall call) {
-    SqlCall substringCall = makeSubstringSqlCall(call);
-    call.setOperand(0, substringCall);
-    if (call.operandCount() == 5 && call.operand(4).toString().equals("'i'")) {
+    if (call.operandCount() == 5 || call.operand(1).toString().contains("\\")) {
```
[the remark this answers is not in the document]
Comment (ankitaspanse) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:589]: We had discussion for the same so I kept iy as it is

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 2 above)
Comment (ankitaspanse) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: Done with changes.Just check once

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java (hunk 3 above)
Comment (ankitaspanse) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:5870]: Actually 3rd and 4th parameters are of numeric data type at both dialects so test case is failing for same if we pass.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java (hunk 2 above)
Comment (Technay) [reply core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java]: you need to substring while returning. I don't see any code changes from previous version

**Review by Technay — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 5]
```diff
@@ -583,51 +581,36 @@ private SqlNode makCastCallForTimeWithPrecision(SqlNode operandToCast, int preci
...
-    String regexStr = call.operand(1).toString();
-    regexStr = regexStr.replace("\\", "\\\\");
-    String regexLiteral = "(?i)".concat(regexStr.substring(1, regexStr.length() - 1));
-    return SqlLiteral.createCharString(regexLiteral,
+    String regexLiteral = call.operand(1).toString();
+    regexLiteral = regexLiteral.replace("\\", "\\\\");
+    if (call.operandCount() == 5 && call.operand(4).toString().equals("'i'")) {
+      regexLiteral = "(?i)".concat(regexLiteral.substring(1, regexLiteral.length() - 1));
+    }
+    return SqlLiteral.createCharString((call.operandCount() == 5) ? regexLiteral
```
Comment (Technay) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:609]: Why is this condition needed. I don't think we discussed this?

**Comment by Technay** [comment]
LGTM

**Review by Technay — APPROVED** [review]

**Pull request merged**